### ISSUE TITLE
Core support for REST API

### ISSUE THREAD

**Issue by LegitTalon** [keystonejs/keystone#396]
[This module](https://www.npmjs.org/package/keystone-rest) is nice but perhaps a RESTful service would be a great addition to Keystone Core.

**Comment by creynders** [comment]
@JedWatson @LegitTalon I wrote a separate module [restful-keystone](https://github.com/d-pac/restful-keystone) which automatically enables a REST API based on your list configuration (and the configuration of restful-keystone) I know alternatives exist, but IMO they want to do too much and don't allow enough granular control.

It's very easy to setup. Assuming you've got a `Article` list in `models`:

```js
// file: routes/index.js
var keystone = require('keystone');

// Pass your keystone instance to the module
var restful = require('restful-keystone')(keystone);

// ...

exports = module.exports = function( app ){
  //Explicitly define which lists we want exposed
  restful.expose({
    Article : true
  }).start();
}
```
```sh
GET /api/articles
```
```sh
Status: 200 OK
{
    "articles": [
        {
            "_id": "54e2f3d7a21780d7a097ce8d",
            "title": "Lorem Ipsum",
            "author": null,
            "categories": [],
            "content": {
                "brief": "<p>Lorem Ipsum</p>",
                "extended": "<p>Dolor sit amet</p>"
            },
            "state": "draft"
        },
        {
            "_id": "54e2f411a21780d7a097ce8e",
            "title": "Just a small Test",
            "publishedDate": "2015-02-16T23:00:00.000Z",
            "author": null,
            "categories": [],
            "content": {
                "brief": "<p>This is a test</p>",
                "extended": "<p>To make sure restful-keystone is functioning correctly</p>"
            },
            "state": "published"
        }
    ]
}
```

The main thing here is that everything is configurable, by default responses are enveloped in an object, but you can turn that off. You can define which methods are allowed on your list, which fields should be shown, editable, populated et cetera. [See the documentation for full configuration details](https://github.com/d-pac/restful-keystone)

I've written a large number of tests, but will need to add some more. Also it's still missing pagination, but it's coming soon, since I need this for current project at work.

Please let me know what you guys think. If you don't like the API, no problem, but if it doesn't function or you find bugs **please give me a shout**, i.e. don't just drop it and leave me oblivious :)

**Comment by sylvainlap** [comment]
What is the status of this request ? Will Keystone add a REST API as a core feature ?

Regards,
Sylvain

**Comment by johannesjo** [comment]
I was wondering the same.

**Comment by jjhesk** [comment]
@creynders  i have done some enhancement of your library and here this is working [new lib](https://github.com/ONode/restful-keystone)

**Comment by MikelArnaiz** [comment]
+1.

**Comment by mxstbr** [comment]
We have the API done and are working on moving the admin interface over to using it completely. This will be released in 0.4!

*See #2566!*
